app/ante/eip712.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	// stdSigs contains the sequence number, account number, and signatures.
+	// When simulating, this would just be a 0-length slice.
+	sigs, err := sigTx.GetSignaturesV2()
+	if err != nil {
+		return ctx, err
+	}
+
+	signerAddrs := sigTx.GetSigners()
+
+	// EIP712 allows just one signature
+	if len(sigs) != 1 {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "invalid number of signer;  EIP712 signatures allows just one signature")
+	}
+
+	// check that signer length and signature length are the same
+	if len(sigs) != len(signerAddrs) {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "invalid number of signer;  expected: %d, got %d", len(signerAddrs), len(sigs))
+	}
+
+	// EIP712 has just one signature, avoid looping here and only read index 0
+	i := 0
+	sig := sigs[i]
+
+	acc, err := authante.GetSignerAcc(ctx, svd.ak, signerAddrs[i])
+	if err != nil {
+		return ctx, err
+	}
+
+	// retrieve pubkey
+	pubKey := acc.GetPubKey()
+	if !simulate && pubKey == nil {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidPubKey, "pubkey on account is not set")
+	}
+
+	// Check account sequence number.
+	if sig.Sequence != acc.GetSequence() {
+		return ctx, sdkerrors.Wrapf(
+			sdkerrors.ErrWrongSequence,
+			"account sequence mismatch, expected %d, got %d", acc.GetSequence(), sig.Sequence,
+		)
+	}
+
+	// retrieve signer data
+	genesis := ctx.BlockHeight() == 0
+	chainID := ctx.ChainID()
+
+	var accNum uint64
+	if !genesis {
+		accNum = acc.GetAccountNumber()
+	}
+
+	signerData := authsigning.SignerData{
+		ChainID:       chainID,
+		AccountNumber: accNum,
+		Sequence:      acc.GetSequence(),
+	}
+
+	if simulate {
+		return next(ctx, tx, simulate)
+	}
+
+	authSignTx, ok := tx.(authsigning.Tx)
+	if !ok {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "tx %T doesn't implement the authsigning.Tx interface", tx)
+	}
+
+	if err := VerifySignature(pubKey, signerData, sig.Data, svd.signModeHandler, authSignTx); err != nil {
+		errMsg := fmt.Errorf("signature verification failed; please verify account number (%d) and chain-id (%s): %w", accNum, chainID, err)
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, errMsg.Error())
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// VerifySignature verifies a transaction signature contained in SignatureData abstracting over different signing modes
+// and single vs multi-signatures.
+func VerifySignature(
+	pubKey cryptotypes.PubKey,
+	signerData authsigning.SignerData,
+	sigData signing.SignatureData,
+	_ authsigning.SignModeHandler,
+	tx authsigning.Tx,
+) error {
+	switch data := sigData.(type) {
+	case *signing.SingleSignatureData:
+		if data.SignMode != signing.SignMode_SIGN_MODE_LEGACY_AMINO_JSON {
+			return fmt.Errorf("unexpected SignatureData %T: wrong SignMode", sigData)
+		}
+
+		// Note: this prevents the user from sending thrash data in the signature field
+		if len(data.Signature) != 0 {
```

Comment:
Lets force the signature data on the cosmos transaction to be empty, so users can not send random data there